### ISSUE TITLE
Simplify inter-node communication interfaces

### ISSUE THREAD

**Issue by jaffee** [pilosa/pilosa#1247]
### Description

Pilosa has 3 interfaces for setting up inter-node communication - you'll notice that two are very similar:
```go
// EventReceiver is the interface for the object which will listen for and
// decode broadcast messages before passing them to pilosa to handle. The
// implementation of this could be an http server which listens for messages,
// gets the protobuf payload, and then passes it to
// EventHandler.ReceiveMessage.
type EventReceiver interface {
	// Start starts listening for broadcast messages - it should return
	// immediately, spawning a goroutine if necessary.
	Start(EventHandler) error
}

// MemberSet represents an interface for Node membership and inter-node communication.
type MemberSet interface {
	// Open starts any network activity implemented by the MemberSet
	// Node is the local node, used for membership broadcasts.
	Open(n *Node) error
}

// BroadcastReceiver is the interface for the object which will listen for and
// decode broadcast messages before passing them to pilosa to handle. The
// implementation of this could be an http server which listens for messages,
// gets the protobuf payload, and then passes it to
// BroadcastHandler.ReceiveMessage.
type BroadcastReceiver interface {
	// Start starts listening for broadcast messages - it should return
	// immediately, spawning a goroutine if necessary.
	Start(BroadcastHandler) error
}
```

Pilosa currently defines several interfaces for sending, and receiving inter-node communication. These are:

```go
// BroadcastHandler is the interface for the pilosa object which knows how to
// handle broadcast messages. (Hint: this is implemented by pilosa.Server)
type BroadcastHandler interface {
	ReceiveMessage(pb proto.Message) error
}

// EventHandler is the interface for the pilosa object which knows how to
// handle broadcast messages. (Hint: this is implemented by pilosa.Server)
type EventHandler interface {
	ReceiveEvent(e *NodeEvent) error
}

// StatusHandler specifies the methods which an object must implement to share
// state in the cluster. These are used by the GossipMemberSet to implement the
// LocalState and MergeRemoteState methods of memberlist.Delegate
type StatusHandler interface {
	LocalStatus() (proto.Message, error)
	ClusterStatus() (proto.Message, error)
	HandleRemoteStatus(proto.Message) error
}

// Gossiper is an interface for sharing messages via gossip.
type Gossiper interface {
	SendAsync(pb proto.Message) error
}

// Broadcaster is an interface for broadcasting messages.
type Broadcaster interface {
	SendSync(pb proto.Message) error
	SendAsync(pb proto.Message) error
	SendTo(to *Node, pb proto.Message) error
}

// InternalClient should be implemented by any struct that enables any transport between nodes
type InternalClient interface {
    // Very long, here's a summary
    // Schema stuff get, create, delete, etc.
    // Import, Export, Backup, Restore
    // FragmentBlocks, BlockData, AttrDiff
    // Query, MaxSlice
    SendMessage(ctx context.Context, pb proto.Message)
}

```

We can roughly categorize these into sending and receiving:
```go
// Sending
StatusHandler
	LocalStatus() (proto.Message, error)
	ClusterStatus() (proto.Message, error)
Gossiper
	SendAsync(pb proto.Message) error
Broadcaster 
	SendSync(pb proto.Message) error
	SendAsync(pb proto.Message) error
	SendTo(to *Node, pb proto.Message) error
InternalClient
  SendMessage(ctx context.Context, pb proto.Message)
  ...

// Receiving
BroadcastHandler
	ReceiveMessage(pb proto.Message) error
EventHandler 
	ReceiveEvent(e *NodeEvent) error
StatusHandler
	HandleRemoteStatus(proto.Message) error
```

Confusingly, there are both pull and push based interfaces here.
Gossip/memberlist has a facility for sharing state between nodes asynchronously
that it polls on its own time (the LocalStatus() and ClusterStatus() methods).
I'm not convinced that it's a good idea for us to use this - it might be better
to use SendAsync or SendSync for consistency.

Currently, Pilosa has two types of communication:

1. A handler (now HTTP), which has some endpoints that are purely external, some
   which are internal, and some which can be either.

2. Gossip/memberlist which is used purely for internal communication. Within
   this, there are cluster membership events (EventHandler), state sharing
   events (StatusHandler), and generic messaging (BroadcastHandler).

It's clear to me that this all can be simplified, but there are a number of
considerations. 

At a high level, Pilosa should expose the operations that can be done to it, and
the information that can be gotten from it, this is mostly in API right now, but
not completely. Implementations of external interfaces (HTTP, GRPC, etc.) can
take an API and expose it however they like.

 Pilosa should define an interface for talking to other nodes. As part of its
initialization, it needs to be given an implementation of this interface. It is
up to the operator of the cluster to ensure that the sender on each node is
compatible with a receiver on every other node. This interface is mostly
InternalClient, but there is a lot of other messiness.

### Success criteria (What criteria will consider this ticket closeable?)

This ticket defines the desired end state at a high level. (one way to send, one
way to receive). Work toward this goal will need to be broken out into subtasks.

**Comment by travisturner** [comment]
Regarding gossip/memberlist, we currently only use it for two things:

1. To determine membership, and more specifically to receive `NodeJoin` events
2. For broadcasting messages asynchronously to the cluster. We stopped using memberlist for `SendSync` messages, and the only place `SendAsync` is used anymore is when a node creates a new slice and that needs to be communicated to the rest of the cluster. I think it would be reasonable to stop using memberlist for this one message.

So in short, I think we could limit memberlist's role to providing `NodeJoin` (and potentially `NodeLeave`) events.

**Comment by jaffee** [comment]
@travisturner what about the cluster state/status stuff?

**Comment by travisturner** [comment]
@jaffee no, the coordinator sends `ClusterStatus` with `SendSync`. We do use the `NodeMeta` information that gets passed along with the `NodeJoin` event, but not anywhere else.

**Comment by jaffee** [comment]
The `pilosa/gossip` package has references to pilosa.StatusHandler as part of implementing memberlist.Delegate. Can we just rip that out?

**Comment by travisturner** [comment]
@jaffee I think it would be safe to remove the `StatusHandler`. Really its only function is to act as a backup for keeping `maxSlice` and `schema` in sync among nodes in the cluster. I say backup, because `maxSlice` and `schema` should be kept up to date through normal `SendSync` messages as they change, but in the case where something gets out of sync, the `StatusHandler` will periodically fix it.

**Comment by jaffee** [comment]
so @travisturner you think it's reasonable to use SendSync in View.createFragmentIfNotExists?

to me this seems like the much safer way to go - we can guarantee that the cluster is in agreement about what the max slice is when setting a bit rather than just hoping everything stays consistent. There will likely be something of a performance penalty when doing high rate imports that are creating lots of slices, but I imagine we can cross that bridge later on. 

If we remove that usage of Async, we can remove the Broadcaster functionality from gossip entirely which should make it easier to disentangle more of this stuff.

**Comment by travisturner** [comment]
@jaffee I think that's a reasonable thing to do.
